ast/src/parsed/mod.rs

```diff
@@ -332,29 +333,163 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
-    Reference(Ref),
-    PublicReference(String),
-    // A number literal and its type.
-    Number(#[schemars(skip)] BigUint, Option<Type>),
-    String(String),
-    Tuple(Vec<Self>),
-    LambdaExpression(LambdaExpression<Self>),
-    ArrayLiteral(ArrayLiteral<Self>),
-    BinaryOperation(Box<Self>, BinaryOperator, Box<Self>),
-    UnaryOperation(UnaryOperator, Box<Self>),
-    IndexAccess(IndexAccess<Self>),
-    FunctionCall(FunctionCall<Self>),
-    FreeInput(Box<Self>),
-    MatchExpression(Box<Self>, Vec<MatchArm<Self>>),
-    IfExpression(IfExpression<Self>),
-    BlockExpression(Vec<StatementInsideBlock<Self>>, Box<Self>),
+    Reference(SourceRef, Ref),
+    PublicReference(SourceRef, String),
+    Number(SourceRef, Number),
+    String(SourceRef, String),
+    Tuple(SourceRef, Vec<Self>),
+    LambdaExpression(SourceRef, LambdaExpression<Self>),
+    ArrayLiteral(SourceRef, ArrayLiteral<Self>),
+    BinaryOperation(SourceRef, BinaryOperation<Self>),
+    UnaryOperation(SourceRef, UnaryOperation<Self>),
+    IndexAccess(SourceRef, IndexAccess<Self>),
+    FunctionCall(SourceRef, FunctionCall<Self>),
+    FreeInput(SourceRef, Box<Self>),
+    MatchExpression(SourceRef, MatchExpression<Self>),
+    IfExpression(SourceRef, IfExpression<Self>),
+    BlockExpression(SourceRef, BlockExpression<Self>),
+}
+
+// Comparison function for expressions that ignore source information.
+impl<Ref: PartialEq> PartialEq for Expression<Ref> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Expression::Reference(_, a), Expression::Reference(_, b)) => a == b,
+            (Expression::PublicReference(_, a), Expression::PublicReference(_, b)) => a == b,
+            (Expression::Number(_, a), Expression::Number(_, b)) => a == b,
+            (Expression::String(_, a), Expression::String(_, b)) => a == b,
+            (Expression::Tuple(_, a), Expression::Tuple(_, b)) => a == b,
+            (Expression::LambdaExpression(_, a), Expression::LambdaExpression(_, b)) => a == b,
+            (Expression::ArrayLiteral(_, a), Expression::ArrayLiteral(_, b)) => a == b,
+            (Expression::BinaryOperation(_, a), Expression::BinaryOperation(_, b)) => a == b,
+            (Expression::UnaryOperation(_, a), Expression::UnaryOperation(_, b)) => a == b,
+            (Expression::IndexAccess(_, a), Expression::IndexAccess(_, b)) => a == b,
+            (Expression::FunctionCall(_, a), Expression::FunctionCall(_, b)) => a == b,
+            (Expression::FreeInput(_, a), Expression::FreeInput(_, b)) => a == b,
+            (Expression::MatchExpression(_, a), Expression::MatchExpression(_, b)) => a == b,
+            (Expression::IfExpression(_, a), Expression::IfExpression(_, b)) => a == b,
+            (Expression::BlockExpression(_, a), Expression::BlockExpression(_, b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+pub trait SourceInfo {
+    fn get_source(&self) -> &SourceRef;
```

Comment:
Shouldn't this just be `source(&self)`? Also it's not really the source, it's a source reference, but maybe we can drop that distinction.